DataFormats/L1Scouting/interface/OrbitCollection.h

```diff
@@ -0,0 +1,186 @@
+
+#include "DataFormats/Common/interface/FillView.h"
+#include "DataFormats/Common/interface/fillPtrVector.h"
+#include "DataFormats/Common/interface/setPtr.h"
+#include "DataFormats/Common/interface/traits.h"
+#include "FWCore/Utilities/interface/GCCPrerequisite.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include <cstdint>
+#include <vector>
+template <class T>
+class OrbitCollection {
+public:
+  typedef typename std::vector<T>::iterator iterator;
+  typedef typename std::vector<T>::const_iterator const_iterator;
+  typedef T value_type;
+  typedef typename std::vector<T>::size_type size_type;
+
+  // initialize the offset vector with 0s from 0 to 3565.
+  // BX range is [1,3564], an extra entry is needed for the offserts of the last BX
+  OrbitCollection() : bxOffsets_(3566, 0), data_(0) {}
+  OrbitCollection(std::vector<std::vector<T>>& orbitBuffer, unsigned nObjects = 0)
+      : bxOffsets_(3566, 0), data_(nObjects) {
+    fillAndClear(orbitBuffer, nObjects);
+  }
+
+  OrbitCollection(const OrbitCollection& other) = default;
+  OrbitCollection(OrbitCollection&& other) = default;
+  OrbitCollection& operator=(const OrbitCollection& other) = default;
+  OrbitCollection& operator=(OrbitCollection&& other) = default;
+
+  void swap(OrbitCollection& other) {
+    using std::swap;
+    swap(bxOffsets_, other.bxOffsets_);
+    swap(data_, other.data_);
+  }
+
+  // Fill the orbit collection starting from a vector of vectors, one per BX.
+  // Objects are moved into a flat data vector and a vector is used to keep track
+  // of the starting offset of every BX.
+  // Input vector must be sorted with increasing BX and contain 3565 elements (BX in [1,3564])
+  void fillAndClear(std::vector<std::vector<T>>& orbitBuffer, unsigned nObjects = 0) {
+    if (orbitBuffer.size() != 3565)
+      throw cms::Exception("OrbitCollection::fillAndClear")
+          << "Trying to fill the collection by passing an orbit buffer with incorrect size. "
+          << "Passed " << orbitBuffer.size() << ", expected 3565" << std::endl;
+    data_.reserve(nObjects);
+    bxOffsets_[0] = 0;
+    unsigned bxIdx = 1;
+    for (auto& bxVec : orbitBuffer) {
+      // increase offset by the currect vec size
+      bxOffsets_[bxIdx] = bxOffsets_[bxIdx - 1] + bxVec.size();
+
+      // if bxVec contains something, move it into the data_ vector
+      // and clear bxVec objects
+      if (bxVec.size() > 0) {
+        data_.insert(data_.end(), std::make_move_iterator(bxVec.begin()), std::make_move_iterator(bxVec.end()));
+        bxVec.clear();
+      }
+
+      // increment bx index
+      bxIdx++;
+    }
+  }
+
+  // iterate over all elements contained in data
+  const_iterator begin() const { return data_.begin(); }
+  const_iterator end() const { return data_.end(); }
+
+  // iterate over elements of a bx
+  const_iterator begin(unsigned bx) const { return bxRange(bx).first; }
+  const_iterator end(unsigned bx) const { return bxRange(bx).second; }
```

Comment:
Using these two functions to iterate over elements of a bx leads to `bxRange()` being called twice, even if in principle one call would suffice. Also the caller needs to repeat the `bx`. How about changing `bxRange()` to return an object that contains the two iterators, and provides access to them via `begin()` and `end()` functions (e.g. [`edm::Span`](https://github.com/cms-sw/cmssw/blob/master/FWCore/Utilities/interface/Span.h)), so that it could be used directly e.g. in range-`for`?

Reply:
Thanks for this suggestion. I've replaced the two existing methods with `edm::Span<const_iterator> bxIterator(unsigned bx)`